src/main/java/uk/gov/nationalarchives/pdi/step/xml/CanonicalStep.java

```diff
@@ -0,0 +1,121 @@
+            data.setXmlFieldIdx(getInputRowMeta().indexOfValue(meta.getInputField()));
+            data.setOutputFieldIndex(data.getOutputRowMeta().indexOfValue(meta.getOutputField()));
+            org.apache.xml.security.Init.init();
+            data.setCanonicalizer(getCanonicalizer());
+            data.setDocumentBuilder(getDocumentBuilder());
+        }
+        final Object xmlFieldValue = row[data.getXmlFieldIdx()];
+        if(xmlFieldValue instanceof String) {
+            final CanonicalizationResult result = process((String)xmlFieldValue,data);
+            if(result.hasErrors()) {
+                putError(data.getOutputRowMeta(), row, result.getErrorCount(), result.getErrorMessage(), meta.getInputField(), "CanonicalStep001");
+            } else {
+                Object[] outputRow = RowDataUtil.resizeArray(row,data.getOutputRowMeta().size());
+                outputRow[data.getOutputFieldIndex()] = result.getCanonicalXml();
+                putRow(data.getOutputRowMeta(),outputRow);
+            }
+            return true;
+        } else {
+            throw new KettleException("Expected field " + meta.getInputField() + " to contain XML as type java.lang.String, but found "
+                    + xmlFieldValue.getClass());
+        }
+    }
+
+    private CanonicalizationResult process(String xmlString, CanonicalStepData data) {
+        try{
+            final Document xmlDoc = createDocument(xmlString,data);
+            return new CanonicalizationResult(canonicalize(xmlDoc,data));
+        } catch (KettleException kex) {
+            CanonicalizationResult result = new CanonicalizationResult("");
+            result.setErrorMessage(kex.getMessage());
+            result.setHasErrors(true);
+            result.incrementErrorCount();
+            return result;
+        }
+    }
+
+    private Canonicalizer getCanonicalizer() throws KettleException {
+        try {
+            return Canonicalizer.getInstance(Canonicalizer.ALGO_ID_C14N_EXCL_WITH_COMMENTS);
+        } catch (Exception e) {
+            throw new KettleException(e.getMessage(),e);
+        }
+    }
+
+    private DocumentBuilder getDocumentBuilder() throws KettleException {
+        final DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+        try {
+            return dbf.newDocumentBuilder();
+        } catch (Exception e) {
+            throw new KettleException(e.getMessage(), e);
+        }
+    }
+
+    private Document createDocument(String xmlString, final CanonicalStepData data) throws KettleException {
+        try{
+            return data.getDocumentBuilder().parse(new InputSource(new StringReader(xmlString)));
+        } catch (Exception e) {
+            throw new KettleException(e.getMessage(),e);
+        }
+    }
+
+    protected static String canonicalize(Document document, CanonicalStepData data) throws KettleException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
Please put the `ByteArrayOutputStream` in the try below as a `try-with-resources`. You can then remove the `IOUtils.closeQuietly`